### ISSUE TITLE
Metrics and autoscaler support

### ISSUE THREAD

**Issue by huoqifeng** [confidential-containers/cloud-api-adaptor#59]
I'm creating this issue to wrap my rough idea for metrics and Pods autoscaler. 

Kubernetes do vertical and horizontal scale based on metrics collected from Node, like based on CPU or Memory, which can also based on `customized metrics`, [horizontal-pod-autoscale](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) explained it clearly. @bpradipt raised [issue](https://github.com/confidential-containers/cloud-api-adaptor/issues/7) and suggest to use `admission mutation webhook` and `advertises extended resources` to do resource management. With this extended resources in place, we can implement a `customized metrics apiserver` and collect these data, so that it can be used by the scaler. An custom-metrics-apiserver example can be found in [custom-metrics-apiserver](https://github.com/kubernetes-sigs/custom-metrics-apiserver). 

The open problem is how to collect metrics data, these data is actually the CPU/Memory used by PeerPod and reflecting to the `advertises extended resources` described in  [issue](https://github.com/confidential-containers/cloud-api-adaptor/issues/7) . I think it can be collected by adaptor in Node or the individual PeerPod. One approach might be as following:

1. Resource Management define `advertises extended resources` and map Pod CPU/Memory resources to that  `advertises extended resources` via webhook
2. Add a custom-metrics-apiserver to collect these `advertises extended resources` as customized-metrics
3. A service in PeerPod report these customized-metrics per Pod to custom-metrics-apiserver
4. Specify the customized-metrics as HorizontalPodAutoscaler of the deployment replicas

**Comment by ariel-adam** [comment]
@huoqifeng is this issue still relevant or can be closed? 
If it's still relevant to what release do you think we should map it to (mid-November, end-December, mid-February etc...)?

**Comment by huoqifeng** [comment]
I'd like set it to `end-December`

**Comment by ariel-adam** [comment]
Added it to the CoCo releases project, drop V0.3.0 priority medium

**Comment by ariel-adam** [comment]
@huoqifeng should this issue be in-progress? 
Do you think it can converge on time for the V0.3.0 release on the 22nd of January?

**Comment by huoqifeng** [comment]
@ariel-adam no, there is no people working on this so far.

**Comment by SamYuan1990** [comment]
https://github.com/kubernetes-sigs/controller-runtime/blob/222fb669e109f4bffdf7fc14a2ac02a2a6e496ba/pkg/manager/manager_test.go#L1187-L1225

found a code sample to create and use Metrics in webhook.

**Comment by bpradipt** [comment]
@SamYuan1990 @huoqifeng  any ideas on this issue ?

**Comment by ariel-adam** [comment]
Given that moving it out of the current release 0.6.0

**Comment by inatatsu** [comment]
@huoqifeng @bpradipt I wonder both the horizontal and vertical autoscalers work as expected with the current mutating webhook solution because the autoscalers:

- rely on the CPU and memory pod metrics which are handled by the kata agent, and
- update the original pod spec which is not yet mutated.

Is my understanding correct?

**Comment by huoqifeng** [comment]
> @huoqifeng @bpradipt I wonder both the horizontal and vertical autoscalers work as expected with the current mutating webhook solution because the autoscalers:
> 
>     * rely on the CPU and memory pod metrics which are handled by the kata agent, and
> 
>     * update the original pod spec which is not yet mutated.
> 
> 
> Is my understanding correct?

I don't think so but @bpradipt might correct me. The mutating webhook solution today can calculate the total virtual CPU/Memories on a node and record what PeerPods consumed and decide whether a new PeerPod can be scheduled to this node. But can not collect/report CPU/Memory metrics that will be used for autoscaling. We'll need a mechanism to collect PeerPods metrics and leverage it for horizontal scaling.

**Comment by inatatsu** [comment]
@huoqifeng The autoscalers ask the metrics server to collect CPU and memory usage of a pod. The request is eventually forwarded to kubelet -> container runtime -> kata-shim -> kata-agent. I think kata-agent uses the cgroups crate and can collect CPU and memory metrics of the containers in a peer pod.
